### ISSUE TITLE
Change ConfigTemplate to use json-schema ?

### ISSUE THREAD

**Issue by escapewindow** [scriptharness/python-scriptharness#28]
http://json-schema.org
http://schemas.taskcluster.net/queue/v1/task.json

I think this will work very well for config definition and validation, and I love that it's a standard.
What we lose is the ability to also define commandline options, essentially everything argparse related.

We may need to define a mix of json-schema and argparse for our uses.

**Comment by carocad** [comment]
Maybe using [docopt](https://github.com/docopt/docopt) can help you have commandline options with config definition and validation in a simple way ;)

**Comment by escapewindow** [comment]
Everyone wants me to use something else for config :)
Others have suggested https://github.com/mozilla/configman

Does docopt support
- non-commandline config definition+validation
- the ability to read in multiple config files and add them to the config ?

At first glance it's looking like only a commandline tool, not config files as well.

**Comment by carocad** [comment]
well after reading a bit about configman, it seems to provide a consistent solution for both config files and command line options, which is not the case for docopt (only command line options)
What I had in mind (since I didnt know such a thing existed) was using the json schema for the config files and docopt for the commandline options: since those become pretty straightforward with docopt, the only thing to code would be the json-schema ;)

**Comment by escapewindow** [comment]
I agree docopt is pretty and simple, but it appears to be missing a few features that scriptharness needs: http://escapewindow.dreamwidth.org/246718.html

**Comment by carocad** [comment]
yeah that's true.
I checked the ideas that you had on your blog and please correct me if I got them wrong, but couldn't this be done by combining docopt, a json parser and json-schema in a straightforward way? something in this direction:

#theoretical file

```
'''
Usage:
  scriptharness [--config <filepath>...]
'''

if __name__ == '__main__':
  # this should load a json file; not a python module (just for consistency)
  config = LogggingDict(load_initial_config())
  validate(config)
  arguments = docopt(__doc__)
  for configfilepath in arguments['filepath']:
    tmpconfig = parsejson(configfilepath)
    config.update(tmpconfig) # always log the changes right?
    validate(config) # increasively validate the full configuration

  run(ReadOnlyDict(config))
```

obviously this will make scriptharness depend on 3 different packages, but those would be very straightforward and easily recognizeble directly from the snippet above; instead of BYOC.
From my point of view, you are aiming at a very specific idea that can only be realized either using different packages or doing it yourself.

I am personally more inclined to use different packages as I would know that each one complies with a standard (posix, json, json-schema); sacrifice code purity for standard compliant.

hope it helps ;)

**Comment by escapewindow** [comment]
With docopt, how do you have 2 or more scripts share a set of commandline options, and get changes automatically when that shared set changes?

**Comment by carocad** [comment]
I am not entirely sure if I understood you correctly, but I will go with my best guess ;)
If what you want if to have two console entry points, e.g:

```
python module1.py --option1 foo --option2 bar
python module2.py --option1 foo --option2 bar
```

then what you should do (afaik) is just share the string passed to docopt. In the snippet from before I used **doc** because the docstring from the module gets assigned to that variable, thus I just retrieved what I wrote before.
Theoretically (I haven't done it), if you could create a string that has the form required by docopt them there shouldnt be any problem. Something like:

```
arguments = docopt(module1.__usage__ + module2.__usage__)
```

**Comment by escapewindow** [comment]
So, for a more real world example:
- all scriptharness scripts use --action, --action-group, --config-file, etc
- Firefox build uses a set of options
- Fennec build (Firefox for Android) uses a similar set of options
- Localization uses a set of options
- Multilocale uses a similar set of options
- Fennec multilocale uses a large set of all of the above options

In ConfigTemplate, it's possible to get the shared set of Firefox + Fennec build options and have them be a single grouping.  Localization and multilocale could also have a single grouping of options.  So, for Fennec multilocale, you would add together the scriptharness options + fennec options (which include the shared firefox options) + multilocale options (which include the shared localization options) + any additional options needed for the specific script.

Docopt __usage__ is a string.  I suppose you could re.sub() or something to insert your options into various parts of the string, but that gets very ugly quickly.  I think one of the advantages of using docopt is its ease of use and pretty output, and I think that goes out the window once you get to this level of complexity.

**Comment by carocad** [comment]
ok, now I fully get what you mean.
You are right, for that level of complexity, the strings at each module would get too messy too easily.
I guess that if you decide for docopt, you would end up with something very similar to ConfigTemplate, i.e. programatically populate the usage string to later assign it as the module usage string.

```
module_usage = get_usage_template()
module_usage.add_option("--new-argument", action='store',
                                            help="help message for --new-argument")
__doc__ = module_usage.to_string()
arguments = docopt(__doc__)
```

I guess it would be a little bit less work adapting docopt to share commandline options than adapting ConfigTemplate but well ... that's just a very wild guess.

**Comment by escapewindow** [comment]
Once you use add_option, how is docopt any better than argparse?

Another couple of issues with docopt sharing commandline options:

 a) when do you add the 'help' option ('-h' and/or '--help')?  If this is in a shared set of options, you may end up with multiple help options
 b) what do you do with conflicting options?  one set has -l for locale, and another has -l for logfile.

If you seriously want to pursue docopt as a solution, it may be useful for us to have working code to look at.  Are you interested in writing a proof of concept here, now that you know the requirements?

**Comment by carocad** [comment]
Hey Aki, I would like to tackle this challenge but I'm still doing the docstrings so I won't have time to do it :/

In response to your questions:
a) I think the help option could be written in the template but overwritten on use
b) I could think of either raise an exception if there are conflicting options or either overwrite them. I personally incline myself for the first one.

Just as a side note: would it not be possible to also define the commandline options through a json-schema? one were you could just put the files to use as templates and define extra options or extra usages?

**Comment by escapewindow** [comment]
Ok.  If/when you start, please be aware that I'm not completely sold on the idea of changing the ConfigTemplate to something else, so any solution we would replace it with would need to be a clear improvement, not just a lateral change.

For (b), in ConfigTemplate I added a remove_option() so you could remove the commandline option from a ConfigVariable.  This works better in ConfigTemplate, where you can have ConfigVariables without any commandline options, though the behavior could make a variable go away if it has no more options left.

As far as defining the options through json-schema, yes, that's one of the options I wrote about in that blog post.  If we extend the schema to include that, then yes, we can use that, but it will no longer be a pure json-schema and I'm not sure if other json-schema implementations will accept our schemas.

There's also various dynamic commandline options, like --action which builds its action list from what the script tells it to.  Adding the actions to the schema would potentially make it less extensible.  There are also validation functions that wouldn't fit in json.  I imagine if we used json-schema we'd use it, plus extend it, and also have some python-side stuff that added dynamic elements outside the schema.
